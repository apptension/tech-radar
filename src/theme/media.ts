import { BaseThemedCssFunction, css, DefaultTheme } from 'styled-components';
import { complement, isNil, reverse } from 'ramda';

export enum Breakpoint {
  MOBILE = 'mobile',
  TABLET = 'tablet',
  DESKTOP = 'desktop',
  DESKTOP_WIDE = 'desktopWide',
  DESKTOP_FULL = 'desktopFull',
}

export const sizes: Record<Breakpoint, number> = {
  [Breakpoint.DESKTOP_FULL]: 1920,
  [Breakpoint.DESKTOP_WIDE]: 1440,
  [Breakpoint.DESKTOP]: 1280,
  [Breakpoint.TABLET]: 768,
  [Breakpoint.MOBILE]: 320,
};

export const sizesOrdered = [
  Breakpoint.MOBILE,
  Breakpoint.TABLET,
  Breakpoint.DESKTOP,
  Breakpoint.DESKTOP_WIDE,
  Breakpoint.DESKTOP_FULL,
] as const;

const getWindowWidth = () => window.innerWidth;

export const getBreakpointMediaQuery = (breakpoint: Breakpoint) => `(min-width: ${sizes[breakpoint]}px)`;

export const media = (breakpoint: Breakpoint, opts: { landscape?: boolean; retina?: boolean } = {}) => {
  return ((styleTemplate: TemplateStringsArray) => {
    const joinQuery = (...queries: (string | null)[]) => queries.filter(complement(isNil)).join(' and ');

    const sizeQuery = `(min-width: ${sizes[breakpoint]}px)`;
    const landscapeQuery = opts.landscape ? '(orientation: landscape)' : null;
    const retinaQueries = opts.retina ? ['(-webkit-min-device-pixel-ratio: 2)', '(min-resolution: 192dpi)'] : null;

    let query = '';
    if (retinaQueries) {
      query = retinaQueries.map(retinaQuery => joinQuery(sizeQuery, landscapeQuery, retinaQuery)).join(', ');
    } else {
      query = joinQuery(sizeQuery, landscapeQuery);
    }

    return css`
      @media ${query} {
        ${css(styleTemplate)}
      }
    `;
  }) as BaseThemedCssFunction<DefaultTheme>;
};

export const isMobile = () => {
  const width = getWindowWidth();
  return width < sizes[Breakpoint.TABLET];
};

export const isTablet = () => {
  const width = getWindowWidth();
  return width >= sizes[Breakpoint.TABLET] && width < sizes[Breakpoint.DESKTOP];
};

export const isDesktop = () => {
  const width = getWindowWidth();
  return width >= sizes[Breakpoint.DESKTOP];
};

export const getActiveBreakpoint = () => {
  let breakpoint: Breakpoint = Breakpoint.MOBILE;
  const config: Record<Breakpoint, Breakpoint> = {
    [Breakpoint.DESKTOP_FULL]: Breakpoint.DESKTOP_FULL,
    [Breakpoint.DESKTOP_WIDE]: Breakpoint.DESKTOP_WIDE,
    [Breakpoint.DESKTOP]: Breakpoint.DESKTOP,
    [Breakpoint.TABLET]: Breakpoint.TABLET,
    [Breakpoint.MOBILE]: Breakpoint.MOBILE,
  };

  sizesOrdered.forEach(size => {
    if (config[size] && window.matchMedia(getBreakpointMediaQuery(size)).matches) {
      breakpoint = config[size];
    }
  });

  return breakpoint;
};

export const responsiveValue = <Value>(defaultValue: Value, config: Partial<Record<Breakpoint, Value>> = {}) => ({ theme }: { theme: DefaultTheme }) => {
  const matchesCurrentBreakpoint = (breakpoint: Breakpoint) => sizesOrdered.indexOf(breakpoint) <= sizesOrdered.indexOf(theme.activeBreakpoint ?? Breakpoint.MOBILE);
  const matchingBreakpoint = reverse(sizesOrdered).find(size => config[size] && matchesCurrentBreakpoint(size));
  return matchingBreakpoint ? config[matchingBreakpoint] : defaultValue;
};
