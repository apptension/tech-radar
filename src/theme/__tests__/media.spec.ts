import { Breakpoint, media, responsiveValue } from '../media';

describe('theme/media', () => {
  describe('responsiveValue', () => {
    const getValue = responsiveValue(100, { [Breakpoint.DESKTOP]: 200 });

    describe('when no breakpoint matches', () => {
      it('should return default value', () => {
        const val = getValue({ theme: { activeBreakpoint: Breakpoint.TABLET } });
        expect(val).toBe(100);
      });
    });

    describe('when specified breakpoint matches', () => {
      it('should return provided breakpoint value', () => {
        const val = getValue({ theme: { activeBreakpoint: Breakpoint.DESKTOP } });
        expect(val).toBe(200);
      });
    });

    describe('when lower than specified breakpoint matches', () => {
      it('should return provided breakpoint value', () => {
        const val = getValue({ theme: { activeBreakpoint: Breakpoint.DESKTOP_FULL } });
        expect(val).toBe(200);
      });
    });
  });

  describe('media()', () => {
    const CSS_CONTENT = 'padding: 10px;';
    const getMedia = (...args: Parameters<typeof media>) => media(...args)`${CSS_CONTENT}`.join('');

    describe('for specified breakpoint', () => {
      it('should return correct css media query', () => {
        expect(getMedia(Breakpoint.MOBILE)).toMatchSnapshot();
        expect(getMedia(Breakpoint.TABLET)).toMatchSnapshot();
        expect(getMedia(Breakpoint.DESKTOP)).toMatchSnapshot();
        expect(getMedia(Breakpoint.DESKTOP_WIDE)).toMatchSnapshot();
        expect(getMedia(Breakpoint.DESKTOP_FULL)).toMatchSnapshot();
      });
    });

    describe('for specified retina breakpoint', () => {
      it('should return correct css media query', () => {
        expect(getMedia(Breakpoint.MOBILE, { retina: true })).toMatchSnapshot();
        expect(getMedia(Breakpoint.TABLET, { retina: true })).toMatchSnapshot();
        expect(getMedia(Breakpoint.DESKTOP, { retina: true })).toMatchSnapshot();
        expect(getMedia(Breakpoint.DESKTOP_WIDE, { retina: true })).toMatchSnapshot();
        expect(getMedia(Breakpoint.DESKTOP_FULL, { retina: true })).toMatchSnapshot();
      });
    });

    describe('for specified landscape breakpoint', () => {
      it('should return correct css media query', () => {
        expect(getMedia(Breakpoint.MOBILE, { landscape: true })).toMatchSnapshot();
        expect(getMedia(Breakpoint.TABLET, { landscape: true })).toMatchSnapshot();
        expect(getMedia(Breakpoint.DESKTOP, { landscape: true })).toMatchSnapshot();
        expect(getMedia(Breakpoint.DESKTOP_WIDE, { landscape: true })).toMatchSnapshot();
        expect(getMedia(Breakpoint.DESKTOP_FULL, { landscape: true })).toMatchSnapshot();
      });
    });

    describe('for specified retina landscape breakpoint', () => {
      it('should return correct css media query', () => {
        expect(getMedia(Breakpoint.MOBILE, { landscape: true, retina: true })).toMatchSnapshot();
        expect(getMedia(Breakpoint.TABLET, { landscape: true, retina: true })).toMatchSnapshot();
        expect(getMedia(Breakpoint.DESKTOP, { landscape: true, retina: true })).toMatchSnapshot();
        expect(getMedia(Breakpoint.DESKTOP_WIDE, { landscape: true, retina: true })).toMatchSnapshot();
        expect(getMedia(Breakpoint.DESKTOP_FULL, { landscape: true, retina: true })).toMatchSnapshot();
      });
    });
  });
});
