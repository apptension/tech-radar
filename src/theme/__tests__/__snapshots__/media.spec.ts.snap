// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`theme/media media() for specified breakpoint should return correct css media query 1`] = `
"
      @media (min-width: 320px) {
        
      }
    "
`;

exports[`theme/media media() for specified breakpoint should return correct css media query 2`] = `
"
      @media (min-width: 768px) {
        
      }
    "
`;

exports[`theme/media media() for specified breakpoint should return correct css media query 3`] = `
"
      @media (min-width: 1280px) {
        
      }
    "
`;

exports[`theme/media media() for specified breakpoint should return correct css media query 4`] = `
"
      @media (min-width: 1440px) {
        
      }
    "
`;

exports[`theme/media media() for specified breakpoint should return correct css media query 5`] = `
"
      @media (min-width: 1920px) {
        
      }
    "
`;

exports[`theme/media media() for specified landscape breakpoint should return correct css media query 1`] = `
"
      @media (min-width: 320px) and (orientation: landscape) {
        
      }
    "
`;

exports[`theme/media media() for specified landscape breakpoint should return correct css media query 2`] = `
"
      @media (min-width: 768px) and (orientation: landscape) {
        
      }
    "
`;

exports[`theme/media media() for specified landscape breakpoint should return correct css media query 3`] = `
"
      @media (min-width: 1280px) and (orientation: landscape) {
        
      }
    "
`;

exports[`theme/media media() for specified landscape breakpoint should return correct css media query 4`] = `
"
      @media (min-width: 1440px) and (orientation: landscape) {
        
      }
    "
`;

exports[`theme/media media() for specified landscape breakpoint should return correct css media query 5`] = `
"
      @media (min-width: 1920px) and (orientation: landscape) {
        
      }
    "
`;

exports[`theme/media media() for specified retina breakpoint should return correct css media query 1`] = `
"
      @media (min-width: 320px) and (-webkit-min-device-pixel-ratio: 2), (min-width: 320px) and (min-resolution: 192dpi) {
        
      }
    "
`;

exports[`theme/media media() for specified retina breakpoint should return correct css media query 2`] = `
"
      @media (min-width: 768px) and (-webkit-min-device-pixel-ratio: 2), (min-width: 768px) and (min-resolution: 192dpi) {
        
      }
    "
`;

exports[`theme/media media() for specified retina breakpoint should return correct css media query 3`] = `
"
      @media (min-width: 1280px) and (-webkit-min-device-pixel-ratio: 2), (min-width: 1280px) and (min-resolution: 192dpi) {
        
      }
    "
`;

exports[`theme/media media() for specified retina breakpoint should return correct css media query 4`] = `
"
      @media (min-width: 1440px) and (-webkit-min-device-pixel-ratio: 2), (min-width: 1440px) and (min-resolution: 192dpi) {
        
      }
    "
`;

exports[`theme/media media() for specified retina breakpoint should return correct css media query 5`] = `
"
      @media (min-width: 1920px) and (-webkit-min-device-pixel-ratio: 2), (min-width: 1920px) and (min-resolution: 192dpi) {
        
      }
    "
`;

exports[`theme/media media() for specified retina landscape breakpoint should return correct css media query 1`] = `
"
      @media (min-width: 320px) and (orientation: landscape) and (-webkit-min-device-pixel-ratio: 2), (min-width: 320px) and (orientation: landscape) and (min-resolution: 192dpi) {
        
      }
    "
`;

exports[`theme/media media() for specified retina landscape breakpoint should return correct css media query 2`] = `
"
      @media (min-width: 768px) and (orientation: landscape) and (-webkit-min-device-pixel-ratio: 2), (min-width: 768px) and (orientation: landscape) and (min-resolution: 192dpi) {
        
      }
    "
`;

exports[`theme/media media() for specified retina landscape breakpoint should return correct css media query 3`] = `
"
      @media (min-width: 1280px) and (orientation: landscape) and (-webkit-min-device-pixel-ratio: 2), (min-width: 1280px) and (orientation: landscape) and (min-resolution: 192dpi) {
        
      }
    "
`;

exports[`theme/media media() for specified retina landscape breakpoint should return correct css media query 4`] = `
"
      @media (min-width: 1440px) and (orientation: landscape) and (-webkit-min-device-pixel-ratio: 2), (min-width: 1440px) and (orientation: landscape) and (min-resolution: 192dpi) {
        
      }
    "
`;

exports[`theme/media media() for specified retina landscape breakpoint should return correct css media query 5`] = `
"
      @media (min-width: 1920px) and (orientation: landscape) and (-webkit-min-device-pixel-ratio: 2), (min-width: 1920px) and (orientation: landscape) and (min-resolution: 192dpi) {
        
      }
    "
`;
